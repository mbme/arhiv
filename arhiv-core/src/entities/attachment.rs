use std::ops::Deref;

use anyhow::*;
use serde::{Deserialize, Serialize};

use super::Document;
use rs_utils::{get_file_hash_sha256, get_file_name};

pub const ATTACHMENT_TYPE: &'static str = "attachment";

pub struct Attachment(Document);

impl Attachment {
    pub fn is_attachment(document: &Document) -> bool {
        document.document_type == ATTACHMENT_TYPE
    }

    pub fn from(document: Document) -> Result<Self> {
        ensure!(
            Attachment::is_attachment(&document),
            "document {} must be an attachment",
            &document.id,
        );

        Ok(Attachment(document))
    }

    pub fn new(file_path: &str) -> Result<Self> {
        let sha256 = get_file_hash_sha256(file_path)?;
        let filename = get_file_name(file_path).to_string();

        let document = Document::new(
            ATTACHMENT_TYPE.to_string(),
            AttachmentInfo { filename, sha256 }.into(),
        );

        Ok(Attachment(document))
    }

    fn get_data(&self) -> AttachmentInfo {
        serde_json::from_value(self.0.data.clone()).expect("must be able to deserialize")
    }

    pub fn get_hash(&self) -> String {
        self.get_data().sha256
    }

    pub fn get_filename(&self) -> String {
        self.get_data().filename
    }
}

impl Deref for Attachment {
    type Target = Document;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl Into<Document> for Attachment {
    fn into(self) -> Document {
        self.0
    }
}

#[derive(Serialize, Deserialize, Debug)]
struct AttachmentInfo {
    pub filename: String,
    pub sha256: String,
}

impl Into<serde_json::Value> for AttachmentInfo {
    fn into(self) -> serde_json::Value {
        serde_json::to_value(self).expect("must be able to serialize")
    }
}
