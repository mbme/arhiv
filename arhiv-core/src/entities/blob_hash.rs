use core::fmt;

use anyhow::*;
use rs_utils::get_file_hash_sha256;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq, Hash)]
pub struct BLOBHash(String);

impl BLOBHash {
    pub fn from_file(file_path: &str) -> Result<Self> {
        let hash = get_file_hash_sha256(&file_path)?;

        Ok(BLOBHash(hash))
    }

    pub fn from_string(hash: impl Into<String>) -> Self {
        BLOBHash(hash.into())
    }
}

impl fmt::Display for BLOBHash {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl AsRef<str> for BLOBHash {
    fn as_ref(&self) -> &str {
        &self.0
    }
}
