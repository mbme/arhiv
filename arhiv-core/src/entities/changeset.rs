use std::fmt;

use serde::{Deserialize, Serialize};

use super::{Document, Id, Revision};

#[derive(Serialize, Deserialize, Debug)]
#[serde(deny_unknown_fields)]
pub struct Changeset {
    pub data_version: u8,
    pub arhiv_id: String,
    pub base_rev: Revision,
    pub documents: Vec<Document>,
}

impl Changeset {
    #[must_use]
    pub fn serialize(&self) -> String {
        serde_json::to_string(self).expect("Failed to serialize Changeset to json")
    }

    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.documents.is_empty()
    }

    #[must_use]
    pub fn contains(&self, id: &Id) -> bool {
        self.documents.iter().any(|document| &document.id == id)
    }
}

impl fmt::Display for Changeset {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "[{} (schema version {}) Changeset: {} documents since {}]",
            self.arhiv_id,
            self.data_version,
            self.documents.len(),
            self.base_rev,
        )
    }
}
