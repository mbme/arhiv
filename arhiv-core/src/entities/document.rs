use std::fmt;

use anyhow::*;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::schema::DataSchema;

use super::{DocumentData, Id, Refs, Revision, SnapshotId};

pub const ERASED_DOCUMENT_TYPE: &str = "";

pub type Timestamp = DateTime<Utc>;

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct Document {
    pub id: Id,
    pub rev: Revision,
    pub prev_rev: Revision,
    pub document_type: String,
    pub snapshot_id: SnapshotId,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub data: DocumentData,
}

impl Document {
    pub fn new(document_type: impl Into<String>) -> Self {
        Document::new_with_data(document_type, DocumentData::new())
    }

    pub fn new_with_data(document_type: impl Into<String>, data: DocumentData) -> Self {
        let now = Utc::now();

        Document {
            id: Id::new(),
            rev: Revision::STAGING,
            prev_rev: Revision::STAGING,
            snapshot_id: SnapshotId::new(),
            document_type: document_type.into(),
            created_at: now,
            updated_at: now,
            data,
        }
    }

    #[must_use]
    pub fn is_erased(&self) -> bool {
        self.document_type == ERASED_DOCUMENT_TYPE
    }

    #[must_use]
    pub fn is_staged(&self) -> bool {
        self.rev == Revision::STAGING
    }

    pub fn extract_refs(&self, schema: &DataSchema) -> Result<Refs> {
        schema.extract_refs(&self.document_type, &self.data)
    }
}

impl std::str::FromStr for Document {
    type Err = anyhow::Error;

    fn from_str(data: &str) -> Result<Document> {
        serde_json::from_str(data).context("Failed to parse document json")
    }
}

impl fmt::Display for Document {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "[Document {}/{} {}]",
            self.document_type, self.id, self.rev,
        )
    }
}
