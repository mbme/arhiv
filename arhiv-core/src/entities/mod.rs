mod attachment;
mod blob_hash;
mod changeset;
mod changeset_response;
mod document;
mod id;
mod revision;
mod snapshot_id;

pub use attachment::{Attachment, ATTACHMENT_HASH_SELECTOR, ATTACHMENT_TYPE};
pub use blob_hash::BLOBHash;
pub use changeset::Changeset;
pub use changeset_response::ChangesetResponse;
pub use document::{Document, Timestamp, TOMBSTONE_TYPE};
pub use id::Id;
pub use revision::Revision;
pub use snapshot_id::SnapshotId;
