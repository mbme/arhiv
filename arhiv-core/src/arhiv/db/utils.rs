use std::convert::TryInto;

use anyhow::*;
use rusqlite::Row;
use serde_json::Value;

use crate::entities::*;

pub fn extract_document(row: &Row) -> Result<Document> {
    Ok(Document {
        id: row.get("id")?,
        rev: row.get("rev")?,
        prev_rev: row.get("prev_rev")?,
        snapshot_id: row.get("snapshot_id")?,
        document_type: row.get("type")?,
        created_at: row.get("created_at")?,
        updated_at: row.get("updated_at")?,
        data: {
            let data: Value = row.get("data")?;

            data.try_into()?
        },
    })
}

pub fn extract_id(row: &Row) -> Result<Id> {
    row.get("id").context(anyhow!("failed to extract id"))
}
