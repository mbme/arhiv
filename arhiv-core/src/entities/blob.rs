use anyhow::*;

use rs_utils::file_exists;

use crate::entities::BLOBId;

pub struct BLOB {
    pub id: BLOBId,
    pub file_path: String,
}

impl BLOB {
    #[must_use]
    pub fn new(id: BLOBId, data_dir: &str) -> Self {
        BLOB {
            file_path: format!("{}/{}", data_dir, id.get_file_name()),
            id,
        }
    }

    pub fn exists(&self) -> Result<bool> {
        file_exists(&self.file_path)
    }
}
