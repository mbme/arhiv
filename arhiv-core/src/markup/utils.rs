use crate::entities::Id;

const LINK_PREFIX: &str = "ref:";

#[must_use]
pub fn extract_id(text: &str) -> Option<Id> {
    if text.starts_with(LINK_PREFIX) && text.len() > LINK_PREFIX.len() {
        let id = text[LINK_PREFIX.len()..].into();

        Some(id)
    } else {
        None
    }
}

#[must_use]
pub fn create_link(url: &str, text: &str) -> String {
    if text.is_empty() {
        format!("<{}>", url)
    } else {
        format!("[{}]({})", text, url)
    }
}

#[must_use]
pub fn create_ref(id: &Id, text: &str) -> String {
    create_link(&format!("{}{}", LINK_PREFIX, id), text)
}
