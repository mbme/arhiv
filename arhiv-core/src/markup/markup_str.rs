use std::convert::From;
use std::{collections::HashSet, ops::Range};

use anyhow::{bail, ensure, Context, Result};
use pulldown_cmark::{
    Alignment, CodeBlockKind, CowStr, Event, HeadingLevel, LinkType, Options, Parser, Tag,
};
use serde::Serialize;

use super::utils::extract_id;
use crate::entities::*;

pub struct MarkupStr<'a>(CowStr<'a>);

impl<'a> MarkupStr<'a> {
    #[must_use]
    pub fn parse(&self) -> Parser {
        let mut options = Options::empty();
        options.insert(Options::ENABLE_STRIKETHROUGH);
        options.insert(Options::ENABLE_TABLES);
        options.insert(Options::ENABLE_TASKLISTS);

        Parser::new_ext(self.0.as_ref(), options)
    }

    #[must_use]
    pub fn extract_refs(&self) -> HashSet<Id> {
        let mut refs: HashSet<Id> = HashSet::new();

        let parser = self.parse();
        for event in parser {
            if let Event::Start(Tag::Link(_, ref destination, _)) = event {
                if let Some(id) = extract_id(destination) {
                    refs.insert(id);
                }
            }
        }

        refs
    }

    #[must_use]
    pub fn preview(&self, lines: usize) -> Self {
        self.0
            .as_ref()
            .lines()
            .take(lines)
            .collect::<Vec<_>>()
            .join("\n")
            .into()
    }

    pub fn get_ast(&self) -> Result<MarkupElement<'_>> {
        let mut stack: Vec<MarkupElement<'_>> = vec![MarkupElement::Document { children: vec![] }];

        for (event, range) in self.parse().into_offset_iter() {
            match event {
                Event::Text(value) => {
                    stack
                        .last_mut()
                        .context("stack must not be empty")?
                        .add_child(MarkupElement::Text { value, range })?;
                }
                Event::Code(value) => {
                    stack
                        .last_mut()
                        .context("stack must not be empty")?
                        .add_child(MarkupElement::Code { value, range })?;
                }
                Event::Html(value) => {
                    stack
                        .last_mut()
                        .context("stack must not be empty")?
                        .add_child(MarkupElement::Html { value, range })?;
                }
                Event::FootnoteReference(label) => {
                    stack
                        .last_mut()
                        .context("stack must not be empty")?
                        .add_child(MarkupElement::FootnoteReference { label, range })?;
                }
                Event::SoftBreak => {
                    stack
                        .last_mut()
                        .context("stack must not be empty")?
                        .add_child(MarkupElement::SoftBreak { range })?;
                }
                Event::HardBreak => {
                    stack
                        .last_mut()
                        .context("stack must not be empty")?
                        .add_child(MarkupElement::HardBreak { range })?;
                }
                Event::Rule => {
                    stack
                        .last_mut()
                        .context("stack must not be empty")?
                        .add_child(MarkupElement::Rule { range })?;
                }
                Event::TaskListMarker(checked) => {
                    stack
                        .last_mut()
                        .context("stack must not be empty")?
                        .add_child(MarkupElement::TaskListMarker { checked, range })?;
                }

                Event::Start(tag) => {
                    let el = match tag {
                        Tag::Paragraph => MarkupElement::Paragraph {
                            children: vec![],
                            range,
                        },
                        Tag::Heading(level, ..) => MarkupElement::Heading {
                            range,
                            level,
                            children: vec![],
                        },
                        Tag::BlockQuote => MarkupElement::BlockQuote {
                            children: vec![],
                            range,
                        },
                        Tag::CodeBlock(kind) => MarkupElement::CodeBlock {
                            range,
                            kind,
                            children: vec![],
                        },
                        Tag::List(first_item_number) => MarkupElement::List {
                            range,
                            first_item_number,
                            children: vec![],
                        },
                        Tag::Item => MarkupElement::ListItem {
                            children: vec![],
                            range,
                        },
                        Tag::FootnoteDefinition(label) => MarkupElement::FootnoteDefinition {
                            range,
                            label,
                            children: vec![],
                        },
                        Tag::Table(alignments) => MarkupElement::Table {
                            range,
                            alignments,
                            children: vec![],
                        },
                        Tag::TableHead => MarkupElement::TableHead {
                            children: vec![],
                            range,
                        },
                        Tag::TableRow => MarkupElement::TableRow {
                            children: vec![],
                            range,
                        },
                        Tag::TableCell => MarkupElement::TableCell {
                            children: vec![],
                            range,
                        },

                        Tag::Emphasis => MarkupElement::Emphasis {
                            children: vec![],
                            range,
                        },
                        Tag::Strong => MarkupElement::Strong {
                            children: vec![],
                            range,
                        },
                        Tag::Strikethrough => MarkupElement::Strikethrough {
                            children: vec![],
                            range,
                        },

                        Tag::Link(link_type, url, title) => MarkupElement::Link {
                            link_type,
                            url,
                            title,
                            children: vec![],
                            range,
                        },

                        Tag::Image(link_type, url, title) => MarkupElement::Image {
                            link_type,
                            url,
                            title,
                            children: vec![],
                            range,
                        },
                    };

                    stack.push(el);
                }

                Event::End(_tag) => {
                    let el = stack.pop().context("stack must not be empty")?;
                    stack
                        .last_mut()
                        .context("stack must not be empty")?
                        .add_child(el)?;
                }
            }
        }

        ensure!(stack.len() == 1, "only Document must be on stack");

        stack.pop().context("Document must be on stack")
    }
}

impl<'a> From<&'a str> for MarkupStr<'a> {
    fn from(value: &'a str) -> Self {
        MarkupStr(CowStr::Borrowed(value))
    }
}

impl<'a> From<String> for MarkupStr<'a> {
    fn from(value: String) -> Self {
        MarkupStr(CowStr::Boxed(value.into_boxed_str()))
    }
}

pub type Children<'a> = Vec<MarkupElement<'a>>;

#[derive(Serialize, Debug)]
#[serde(deny_unknown_fields, tag = "typeName")]
pub enum MarkupElement<'a> {
    Document {
        children: Children<'a>,
    },
    Text {
        range: Range<usize>,
        value: CowStr<'a>,
    },
    Code {
        range: Range<usize>,
        value: CowStr<'a>,
    },
    Html {
        range: Range<usize>,
        value: CowStr<'a>,
    },
    FootnoteReference {
        range: Range<usize>,
        label: CowStr<'a>,
    },
    SoftBreak {
        range: Range<usize>,
    },
    HardBreak {
        range: Range<usize>,
    },
    Rule {
        range: Range<usize>,
    },
    TaskListMarker {
        range: Range<usize>,
        checked: bool,
    },
    Paragraph {
        range: Range<usize>,
        children: Children<'a>,
    },
    Heading {
        range: Range<usize>,
        level: HeadingLevel,
        children: Children<'a>,
    },
    BlockQuote {
        range: Range<usize>,
        children: Children<'a>,
    },
    CodeBlock {
        range: Range<usize>,
        kind: CodeBlockKind<'a>,
        children: Children<'a>,
    },
    List {
        range: Range<usize>,
        first_item_number: Option<u64>,
        children: Children<'a>,
    },
    ListItem {
        range: Range<usize>,
        children: Children<'a>,
    },
    FootnoteDefinition {
        range: Range<usize>,
        label: CowStr<'a>,
        children: Children<'a>,
    },
    Table {
        range: Range<usize>,
        alignments: Vec<Alignment>,
        children: Children<'a>,
    },
    TableHead {
        range: Range<usize>,
        children: Children<'a>, // only TableCell
    },
    TableRow {
        range: Range<usize>,
        children: Children<'a>, // only TableCell
    },
    TableCell {
        range: Range<usize>,
        children: Children<'a>,
    },

    Emphasis {
        range: Range<usize>,
        children: Children<'a>,
    },
    Strong {
        range: Range<usize>,
        children: Children<'a>,
    },
    Strikethrough {
        range: Range<usize>,
        children: Children<'a>,
    },

    Link {
        range: Range<usize>,
        link_type: LinkType,
        url: CowStr<'a>,
        title: CowStr<'a>,
        children: Children<'a>,
    },

    Image {
        range: Range<usize>,
        link_type: LinkType,
        url: CowStr<'a>,
        title: CowStr<'a>,
        children: Children<'a>,
    },
}

impl<'a> MarkupElement<'a> {
    fn get_children_mut(&mut self) -> Option<&mut Children<'a>> {
        match self {
            MarkupElement::Document { children }
            | MarkupElement::Paragraph { children, .. }
            | MarkupElement::Heading { children, .. }
            | MarkupElement::BlockQuote { children, .. }
            | MarkupElement::CodeBlock { children, .. }
            | MarkupElement::List { children, .. }
            | MarkupElement::ListItem { children, .. }
            | MarkupElement::FootnoteDefinition { children, .. }
            | MarkupElement::Table { children, .. }
            | MarkupElement::TableHead { children, .. }
            | MarkupElement::TableRow { children, .. }
            | MarkupElement::TableCell { children, .. }
            | MarkupElement::Emphasis { children, .. }
            | MarkupElement::Strong { children, .. }
            | MarkupElement::Strikethrough { children, .. }
            | MarkupElement::Link { children, .. }
            | MarkupElement::Image { children, .. } => {
                //
                Some(children)
            }

            _ => None,
        }
    }

    fn add_child(&mut self, child: MarkupElement<'a>) -> Result<()> {
        if let Some(children) = self.get_children_mut() {
            children.push(child);

            Ok(())
        } else {
            bail!("can't push child into {:?}", self)
        }
    }
}

#[cfg(test)]
mod tests {
    use serde_json::Value;

    use crate::markup::MarkupStr;

    fn into_ast(value: &str) -> Value {
        let markup = MarkupStr::from(value);

        let ast = markup.get_ast().expect("must get ast");

        serde_json::to_value(ast).expect("ast must serialize")
    }

    #[test]
    fn test_parse_header() {
        let ast = into_ast("# TEST HEADER");

        insta::assert_json_snapshot!(ast);
    }

    #[test]
    fn test_parse_list() {
        let ast = into_ast(
            "
- [ ] test 1
- [+] test 2
- [x] test 3
- test 4
* test 5
",
        );

        insta::assert_json_snapshot!(ast);
    }

    #[test]
    fn test_parse_hr() {
        let ast = into_ast("---");

        insta::assert_json_snapshot!(ast);
    }

    #[test]
    fn test_parse_table() {
        let ast = into_ast(
            "
| Syntax      | Description | Test Text     |
| :---        |    :----:   |          ---: |
| Header      | Title       | Here's this   |
| Paragraph   | Text        | And more      |
",
        );

        insta::assert_json_snapshot!(ast);
    }

    #[test]
    fn test_parse_code_block() {
        let ast = into_ast(
            r#"
```json
{
  "firstName": "John",
  "lastName": "Smith",
  "age": 25
}
```
"#,
        );

        insta::assert_json_snapshot!(ast);
    }

    #[test]
    fn test_parse_link() {
        let ast = into_ast(r#"[test](http://example.com "some title")"#);

        insta::assert_json_snapshot!(ast);
    }

    #[test]
    fn test_parse_autolink() {
        let ast = into_ast(r#"<http://example.com>"#);

        insta::assert_json_snapshot!(ast);
    }

    #[test]
    fn test_parse_image() {
        let ast = into_ast(r#"![test](http://example.com "some title")"#);

        insta::assert_json_snapshot!(ast);
    }
}
