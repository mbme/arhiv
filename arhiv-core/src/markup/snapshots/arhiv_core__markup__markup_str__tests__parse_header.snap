---
source: arhiv-core/src/markup/markup_str.rs
assertion_line: 394
expression: ast
---
{
  "children": [
    {
      "children": [
        {
          "range": {
            "end": 13,
            "start": 2
          },
          "typeName": "Text",
          "value": "TEST HEADER"
        }
      ],
      "level": "H1",
      "range": {
        "end": 13,
        "start": 0
      },
      "typeName": "Heading"
    }
  ],
  "typeName": "Document"
}
